v03_pipeline/lib/misc/lookup.py

```diff
@@ -0,0 +1,234 @@
+        project_families=(
+            {project_guid: [fs.family_guid for fs in ps] for ps in row[0].project_stats}
+            if has_project_stats
+            else hl.empty_dict(hl.tstr, hl.tarray(hl.tstr))
+        ),
+    )
+    return ht.annotate(
+        project_stats=ht.project_stats.map(
+            lambda ps: ps.map(lambda fs: fs.drop('family_guid')),
+        ),
+    )
+
+
+def remove_new_callset_family_guids(
+    ht: hl.Table,
+    project_guid: str,
+    family_guids: hl.SetExpression,
+) -> hl.Table:
+    if project_guid not in hl.eval(ht.globals.project_families):
+        return ht
+    project_i = ht.project_guids.index(project_guid)
+    family_indexes_to_keep = hl.array(
+        hl.enumerate(ht.globals.project_families[project_guid])
+        .filter(lambda item: ~family_guids.contains(item[1]))
+        .map(lambda item: item[0]),
+    )
+    ht = ht.annotate(
+        project_stats=(
+            hl.enumerate(ht.project_stats).starmap(
+                lambda i, fs: (
+                    hl.if_else(
+                        i != project_i,
+                        fs,
+                        family_indexes_to_keep.map(
+                            lambda j: ht.project_stats[i][j],
+                        ),
+                    )
+                ),
+            )
+        ),
+    )
+    return ht.annotate_globals(
+        project_families=hl.dict(
+            ht.project_families.items().map(
+                lambda item: (
+                    hl.if_else(
+                        item[0] != project_guid,
+                        item,
+                        (
+                            item[0],
+                            ht.project_families[project_guid].filter(
+                                lambda family_guid: ~family_guids.contains(family_guid),
+                            ),
+                        ),
+                    )
+                ),
+            ),
+        ),
+    )
+
+
+def join_lookup_hts(
+    ht: hl.Table,
+    callset_ht: hl.Table,
+) -> hl.Table:
+    ht = ht.join(callset_ht, 'outer')
+    project_guid = ht.project_guids_1[0]
+    ht_project_i = ht.project_guids.index(project_guid)
+    ht = ht.select(
+        # We have 6 unique cases here.
+        # 1) The project has not been loaded before, the row is missing
+        #    on the left but present on the right.
+        # 2) The project has not been loaded before, the row is present
+        #    on the left but missing on the right.
+        # 3) The project as not been loaded before, the row is present on
+        #    both the left and right.
+        # 4) The project has been loaded before, the row is missing on the
+        #    left but present on the right.
+        # 5) The project has been loaded before, the row is present on the
+        #    left but missing on the right.
+        # 6) The project has been loaded before, the row is present on both
+        #    the left and right.
```

Comment:
So its not 100% the same but I had to deal with a similar-ish scenario for merging different project hts while merging project tables in seqr and ended up with the following code to join `project_ht` onto `families_ht`. Its much simpler than what you have here, and I think it covers all the same cases you have but its possible theres an edge case that I don't need to handle the way you do
```
entry_type = families_ht.family_entries.dtype.element_type
default=hl.range(num_families_ht_families).map(lambda i: hl.missing(entry_type))
default_1=hl.range(num_project_ht_families).map(lambda i: hl.missing(entry_type))
families_ht = families_ht.join(project_ht, how='outer')
families_ht = families_ht.select_globals(
     family_guids=families_ht.family_guids.extend(families_ht.family_guids_1)
)
families_ht = families_ht.select(
    family_entries=hl.bind(
       lambda a1, a2: a1.extend(a2),
       hl.or_else(families_ht.family_entries, default),
       hl.or_else(families_ht.family_entries_1, default_1),
    ),
)
```

Reply:
yes, that's pretty much what I've got [here](https://github.com/broadinstitute/seqr-loading-pipelines/blob/main/v03_pipeline/lib/misc/family_entries.py#L126-L147) in the entries table.  

The difference is coming from the nesting of the families within the projects here.  With the entries tables, though it's also an array of arrays, the inner array (family) is always complete.  So merging the old families with the new families is either the "extend" or the population with the array of missing families.

Here, we're handling projects being added wholesale to the end of the existing array AND families being added in at the index of the project.  The top three cases in this `case` statement are more or less what we have in the family entries table, and the bottom three cases are for the second.  It was tricky to split into two separate blocks of three cases each (though I tried!).  Maybe worth trying again?

Reply:
no I think this is okay as is, just wanted to make sure we weren't overlooking an obvious improvement